Merl-779: blocks tutorial rewrite (#1316)
wpengine/faustjs: internal/faustjs.org/docs/gutenberg/tutorial/create-a-block-from-wordpress-core.mdx

description: This tutorial will give you some simple steps to follow to create a Block from the official WordPress Blocks list
---

In this tutorial you will install and create a block using `@faustwp/blocks`. It will use data from the official Blocks that are available in the block editor of your WordPress site.

## Prerequisites

You will need to complete both of these guides before you can create blocks with this tutorial. Make sure you have the latest version of `Faust` and `@faustwp/blocks` installed if you have already completed these guides in the past.

1. [Faust Getting Started Guide](/docs/getting-started)
2. [Blocks Getting Started Guide](/docs/gutenberg/getting-started)

## 1. Choose a block to implement

:::note
The Core Editor Blocks contain a list of useful blocks for editing and publishing content. All of the official blocks include a `block.json` with defined attributes, so the `wp-graphql-content-blocks` plugin will automatically expose those as GraphQL fields.
:::

> For this tutorial, we are going to implement the [Code Block](https://wordpress.com/support/wordpress-editor/blocks/code-block/) that renders a formatted block of code.

Reviewer: Docs should only reference wordpress.org.

- wordpress.org (free & open-source)
- wordpress.com (commercial)

```suggestion
For this tutorial, we are going to implement the [Code Block](https://developer.wordpress.org/block-editor/reference-guides/core-blocks/#code) that renders a formatted block of code.
```

<a href="/docs/img/core-code-block-preview.png" target="_blank">
  <img src="/docs/img/core-code-block-preview.png" alt="Code Block Preview" />
</a>

:::note
You can expand the block sidebar options to change some of its styling. For example, you can configure the Color, Dimensions, Border, and Size settings for the Code Block.

Once you get familiar with the block's look and feel, let's explore its attributes from the GraphiQL IDE.
:::

## 3. Review the Block data and attributes in the GraphiQL IDE

:::note

<img src="/docs/img/graphiql-ide.png" alt="GraphiQL IDE" />

The GraphiQL IDE link will show in the Admin Bar of your WordPress instance if you
have properly installed WPGraphQL from the [Getting Started Guide](/docs/getting-started).
:::

### 3.1. Within the GraphiQL IDE, use the Docs to find the `CoreCode` Block to display its type and properties. If you chose a different Block, then search for that type instead.

After searching the Docs for `CoreCode` and clicking on `CoreCodeAttributes`, you will see the following:

<img
  src="/docs/img/core-code-block-ide-attributes.png"
  alt="Code Block Block attributes documentation in GraphiQL IDE"
/>

> ### 3.2. You can use the following query in the GraphiQL IDE to inspect the Block data. Replace `/posts/testing` with the path to your post.

Reviewer: I'm wondering if this title can be reduced in length 🤔

      ]
    }
```

:::note
Notice that with some of the attributes like `backgroundColor`, the editor used the `theme.json` palette slug name instead of the actual value, i.e. `tertiary`. It also stored the custom hardcoded styles in the `style` fields as a JSON string. We will have to handle them appropriately when developing the associated block.

Once you get familiar with how to query the necessary data from that block, let's see how to implement the block in your Faust Headless site.
:::

## 4. Create the initial block and log the attributes

Follow the steps below to create a `CoreCode` block that logs its attributes in the console.

### 4.1 Create a `wp-blocks` folder in your Headless Site root

<img src="/docs/img/wp-blocks-folder.png" alt="wp-blocks folder" />

### 4.2 Create the `CoreCode.js` inside the `wp-blocks` folder

```jsx title="wp-blocks/CoreCode.js"
// import react
import React from 'react';
// create CoreCode component
export default function CoreCode(props) {
  console.log(props.attributes);
  return <div>CoreCode</div>;
}
```

This component logs the block attributes to the console via the `attributes` prop passed through the `props` parameter.

> ### 4.3 Create an `index.js` file inside the `wp-blocks` folder that imports and then exports the `CoreCode` block. You will use this same file to export other blocks as well.

Reviewer: I'm wondering if this title can be reduced in length 🤔

```jsx title="wp-blocks/index.js"
// import CoreCode
import CoreCode from './CoreCode';
// export blocks
export default {
  CoreCode,
};
```

> :::note
> By exporting the block in `index.js` we make it available to the `WordPressBlocksProvider`.
> :::

Reviewer: Docusaurus admonition should have a full line of space between the opening and closing tags to [prevent issues with prettier](https://docusaurus.io/docs/markdown-features/admonitions#usage-with-prettier).

```suggestion
:::note

By exporting the block in `index.js` we make it available to the `WordPressBlocksProvider`.

:::
```

### 4.4 Create the `CoreCode` block fragment

You want to create a fragment that describes the request block fields and attributes. Include all relevant fields for your implementation. The `CoreCode` file so far should look like this:

```jsx title="wp-blocks/CoreCode.js" {4,11-26}

> // import react
> import React from 'react';
> // import graphql from apollo/client
> import { gql } from '@apollo/client';
> // create CoreCode component
> export default function CoreCode(props) {
>   console.log(props.attributes);
>   return <div>CoreCode</div>;
> }
> // add fragment

Reviewer: ```suggestion
import React from 'react';
import { gql } from '@apollo/client';

export default function CoreCode(props) {
  console.log(props.attributes);
  return <div>CoreCode</div>;
}

```

```jsx title="wp-blocks/CoreCode.js"
// import react
import React from 'react';
// import graphql and our getStyles utility
import { gql } from '@apollo/client';
import getStyles from '../utilities/getStyles.js';

export default function CoreCode(props) {
  // set wp-block-code style for the markup to match core WordPress styling
  const className = 'wp-block-code';
  // use getStyles utility to process the props.attributes properly
  const styles = getStyles(props.attributes);
  return (
    // This markup is from the WordPress CoreCode block
    // add the className and styles to the markup
    // add props.attributes?.content to the markup
    <pre className={className} style={styles}>
      <code>{`${props.attributes?.content}`}</code>
    </pre>
  );
}
// add fragment
CoreCode.fragments = {
  key: `CoreCodeBlockFragment`,
  entry: gql`
    fragment CoreCodeBlockFragment on CoreCode {
      attributes {
        borderColor
        backgroundColor
        content
        style
        textColor
        fontSize
        fontFamily
      }
    }
  `,
};
```

> :::note
> Now that you have the styles configured correctly, you can navigate to your Headless Website and check that the block matches the styles in the WordPress editor. Make sure to tweak the block in the WordPress Editor and verify that the changes are being reflected in the Headless Website.
> :::

Reviewer: https://docusaurus.io/docs/markdown-features/admonitions#usage-with-prettier
```suggestion
:::note

Now that you have the styles configured correctly, you can navigate to your Headless Website and check that the block matches the styles in the WordPress editor. Make sure to tweak the block in the WordPress Editor and verify that the changes are being reflected in the Headless Website.

:::
```
